scipy/optimize/_simulated_annealing.py

```diff
@@ -0,0 +1,558 @@
+            visits = np.array([self.visit_fn(
+                temperature) for _ in range(dim)])
+            upper_sample = self.rand_state.random_sample()
+            lower_sample = self.rand_state.random_sample()
+            visits[visits > self.TAIL_LIMIT] = self.TAIL_LIMIT * upper_sample
+            visits[visits < -self.TAIL_LIMIT] = -self.TAIL_LIMIT * lower_sample
+            x_visit = visits + x
+            a = x_visit - self.lower
+            b = np.fmod(a, self.bound_range) + self.bound_range
+            x_visit = np.fmod(b, self.bound_range) + self.lower
+            x_visit[np.fabs(
+                x_visit - self.lower) < self.MIN_VISIT_BOUND] += 1.e-10
+        else:
+            # Changing only one coordinate at a time based on Markov chain step
+            x_visit = np.copy(x)
+            visit = self.visit_fn(temperature)
+            if visit > self.TAIL_LIMIT:
+                visit = self.TAIL_LIMIT * self.rand_state.random_sample()
+            elif visit < -self.TAIL_LIMIT:
+                visit = -self.TAIL_LIMIT * self.rand_state.random_sample()
+            index = step - dim
+            x_visit[index] = visit + x[index]
+            a = x_visit[index] - self.lower[index]
+            b = np.fmod(a, self.bound_range[index]) + self.bound_range[index]
+            x_visit[index] = np.fmod(b, self.bound_range[
+                index]) + self.lower[index]
+            if np.fabs(x_visit[index] - self.lower[
+                    index]) < self.MIN_VISIT_BOUND:
+                x_visit[index] += self.MIN_VISIT_BOUND
+        return x_visit
+
+    def visit_fn(self, temperature):
+        # Formula Visita from p. 405 of reference [2]
+        factor1 = np.exp(np.log(temperature) / (self.visiting_param - 1.0))
+        factor2 = np.exp((4.0 - self.visiting_param) * np.log(
+            self.visiting_param - 1.0))
+        factor3 = np.exp((2.0 - self.visiting_param) * np.log(2.0) / (
+            self.visiting_param - 1.0))
+        factor4 = np.sqrt(np.pi) * factor1 * factor2 / (factor3 * (
+            3.0 - self.visiting_param))
+        factor5 = 1.0 / (self.visiting_param - 1.0) - 0.5
+        d1 = 2.0 - factor5
+        factor6 = np.pi * (1.0 - factor5) / np.sin(
+            np.pi * (1.0 - factor5)) / np.exp(gammaln(d1))
+        sigmax = np.exp(-(self.visiting_param - 1.0) * np.log(
+            factor6 / factor4) / (3.0 - self.visiting_param))
+        x = sigmax * self.gaussian_fn(1)
+        y = self.gaussian_fn(0)
+        den = np.exp(
+            (self.visiting_param - 1.0) * np.log((np.fabs(y))) / (
+                3.0 - self.visiting_param))
+        return x / den
+
+    def gaussian_fn(self, axis):
+        if axis == 1:
+            enter = True
+            while enter or (self.s_gauss <= 0 or self.s_gauss >= 1):
+                enter = False
+                sample1 = self.rand_state.random_sample()
+                self.x_gauss = sample1 * 2.0 - 1.0
```

Comment:
If these numbers are from a paper as well. I think you can add a reference here too.